frame/support/src/migrations.rs

```diff
@@ -15,19 +15,126 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#[cfg(feature = "try-runtime")]
-use crate::storage::unhashed::contains_prefixed_key;
 use crate::{
-	traits::{GetStorageVersion, NoStorageVersionSet, PalletInfoAccess, StorageVersion},
+	traits::{
+		GetStorageVersion, NoStorageVersionSet, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion,
+	},
 	weights::{RuntimeDbWeight, Weight},
 };
 use impl_trait_for_tuples::impl_for_tuples;
 use sp_core::Get;
 use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
 use sp_std::marker::PhantomData;
+
 #[cfg(feature = "try-runtime")]
 use sp_std::vec::Vec;
 
+/// Make it easier to write versioned runtime upgrades.
+///
+/// VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking
+/// and setting storage versions. Instead, the developer wraps their migration in this struct which
+/// takes care of version handling using best practices.
+///
+/// It takes 4 type parameters:
+/// - `Version`: The version being upgraded *from*.
+/// - `Inner`: An implementation of `OnRuntimeUpgrade`.
+/// - `Pallet`: The Pallet being upgraded.
```

Comment:
The main usage for this will be pallets, so it is probably find to design it pallet-centric.  
I am just thinking if we also want to use this for non-pallet migrations :thinking:

Reply:
Currently, versions are always relative to the on-chain pallet version. Are you thinking we could create on-chain versions for other arbitrary storage groups that we manage with runtime upgrades? Do you have an example?

Reply:
I was just thinking in the general spirit of "storage migrations". These are not necessarily related to pallets - but for our use cases they almost always are.  
Not sure if we can create a `GetVersion` trait or something that can be blanket implemented for the `PalletInfoAccess` trait or something to allow for non-pallet migrations, but it is probably fine for now. As we want to get this shipped :laughing: